### ISSUE TITLE
Limit and Offset support

### ISSUE THREAD

**Issue by teikitel** [jruesga/phoenix-hibernate-dialect#1]
Hello, I would suggest changing the limit handler to also override methods : 

* `useMaxForLimit()` => true
* `bindLimitParametersInReverseOrder()` => true

Because unlike general SQL dialects, OFFSET statement is applied after LIMIT statement. Example : 

* Phoenix SQL : `SELECT * FROM PERSON ORDER BY PERSON.NAME LIMIT 20 OFFSET 10;` => 10 last persons of the 20 first persons. 
* Postgres SQL : `SELECT * FROM PERSON ORDER BY PERSON.NAME LIMIT 20 OFFSET 10;` => 20 following persons after the 10 first persons.

**Comment by jruesga** [comment]
I check the behaviour you described and seems right. "LIMIT 20 OFFSET 10" returns the 20 following persons after the 10 first persons. Running a simple query, match what specification says (https://phoenix.apache.org/paged.html)

"OFFSET 10 LIMIT 20" raises an error and change the order parameter a user input (LIMIT 10 OFFSET 20) to match what you are requesting doesn't seem a good behaviour imho.

**Comment by teikitel** [comment]
Weird, I just checked with this query `SELECT * FROM PERSON ORDER BY PERSON.NAME LIMIT 180 OFFSET 160;` as I described my resultset contains persons from 161 to 180. I am using Phoenix 4.7.

This is how I overrided your Driver : 
```public class PhoenixDialect extends com.ruesga.phoenix.dialect.PhoenixDialect {

    private final Logger log = LoggerFactory.getLogger(this.getClass());

    @Override
    public LimitHandler getLimitHandler() {
        return new AbstractLimitHandler() {
            @Override
            public boolean supportsLimit() {
                return true;
            }

            @Override
            public String processSql(String sql, RowSelection selection) {
                if (LimitHelper.useLimit(this, selection)) {
                    final boolean hasMaxRows = LimitHelper.hasMaxRows(selection);
                    final boolean hasOffset = LimitHelper.hasFirstRow(selection);
                    return sql + (hasMaxRows ? " limit ?" : "")
                        + (hasOffset ? " offset ?" : "");
                }
                return sql;
            }

            @Override
            public boolean useMaxForLimit() {
                return true;
            }

            @Override
            public boolean bindLimitParametersInReverseOrder() {
                return true;
            }
        };
    }
}
```

Just for the record I perfomed a Spring Boot with Spring Data integration of your Driver. Maybe you could add a note in your README.md for Spring Boot integration :

Add the datasource in application.yml : 
```
spring:
    datasource:
        driver-class-name: org.apache.phoenix.jdbc.PhoenixDriver
        url: jdbc:phoenix:URL
        jpa:
            properties:
                hibernate.ddl-auto: none
                hibernate.dialect: com.ruesga.phoenix.dialect.PhoenixDialect
```
As documented [here](https://github.com/subes/invesdwin-instrument)
With spring-boot you have to ensure that the context in the aspects is updated by importing the one responsible for initializing the load time weaving into the spring-boot configuration:

```
@SpringBootApplication
@ImportResource(locations = "classpath:/META-INF/ctx.spring.weaving.xml") //make @Configurable work
public class MySpringBootApplication {
    public static void main(final String[] args) {
        DynamicInstrumentationLoader.waitForInitialized(); //dynamically attach java agent to jvm if not already present
        DynamicInstrumentationLoader.initLoadTimeWeavingContext(); //weave all classes before they are loaded as beans
        SpringApplication.run(MySpringBootApplication.class, args); //start application, load some classes
    }
}
```

Thanks for your great work on this
Regards

**Comment by teikitel** [comment]
Just found out the [source of my problem](https://issues.apache.org/jira/browse/PHOENIX-3342?jql=project%20%3D%20PHOENIX%20AND%20text%20~%20%22limit%20offset%22)
With ORDER BY : `SELECT * FROM PERSON ORDER BY PERSON.NAME LIMIT 20 OFFSET 10;` => 10 last persons of the 20 first persons.
Without ORDER BY : `SELECT * FROM PERSON ORDER BY PERSON.NAME LIMIT 20 OFFSET 10;` => 20 following persons after the 10 first persons.

Seems like it has been fixed since version 4.8.2.

**Comment by jruesga** [comment]
Thanks for the SpringBoot heads up. I will try to update the README.

I tested with version 4.11, so I didn't found that bug, sorry.
